solutions/rust/05-echo/code/src/resp.rs

```diff
@@ -0,0 +1,209 @@
+use anyhow::{Error, Result};
+use bytes::BytesMut;
+use tokio::{io::AsyncReadExt, io::AsyncWriteExt, net::TcpStream};
+
+const CARRIAGE_RETURN: u8 = '\r' as u8;
+const NEWLINE: u8 = '\n' as u8;
+
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub enum Value {
+    /// For Simple Strings the first byte of the reply is "+".
+    String(String),
+    /// For Errors the first byte of the reply is "-".
+    Error(String),
+    /// For Bulk Strings the first byte of the reply is "$".
+    Bulk(String),
+    /// For Arrays the first byte of the reply is "*".
+    Array(Vec<Value>),
+}
+
+impl Value {
+    pub fn to_command(&self) -> Result<(String, Vec<Value>)> {
+        match self {
+            Value::Array(items) => {
+                return Ok((
+                    items.first().unwrap().unwrap_bulk(),
+                    items.clone().into_iter().skip(1).collect(),
+                ));
+            }
+            _ => Err(Error::msg("not an array")),
+        }
+    }
+
+    fn unwrap_bulk(&self) -> String {
+        match self {
+            Value::Bulk(str) => str.clone(),
+            _ => panic!("not a bulk string"),
+        }
+    }
+
+    pub fn encode(self) -> String {
+        match &self {
+            Value::String(s) => format!("+{}\r\n", s.as_str()),
+            Value::Error(msg) => format!("-{}\r\n", msg.as_str()),
+            Value::Bulk(s) => format!("${}\r\n{}\r\n", s.chars().count(), s),
+            // The other cases are not required
+            _ => panic!("value encode not implemented for: {:?}", self),
+        }
+    }
+}
+
+pub struct RespConnection {
+    stream: TcpStream,
+    buffer: BytesMut,
+}
+
+impl RespConnection {
+    pub fn new(stream: TcpStream) -> Self {
+        return RespConnection {
+            stream,
+            buffer: BytesMut::with_capacity(512),
```

Comment:
In the previous stages, we were using `let mut buf = [0; 512];` - and here we're using `BytesMut`. Is it possible to use `BytesMut` in the previous stages too? 

Just to share some context here - we have users with varying levels of skills who read these solutions. Every single "change" from the previous solution can trigger a "Why was this change made?" question, and if these questions aren't easy to answer it can be frustrating for the user.